### PR TITLE
Add websocket chat support

### PR DESCRIPTION
**Pull request by notbakaneko**
Replaces polling for messages and channel presence with broadcasting messages via websockets.

This initial version uses the existing notification server to relay messages to the web clients.

- [x] filter active users per-channel
- [x] https://github.com/ppy/osu-notification-server/pull/41 (just extra support, not actually needed to function)

### CODE REVIEW

**Comment by nanaya** [comment]
Merge conflicts

File: app/Events/ChatMessageEvent.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace App\Events;
+
+use App\Models\Chat\Message;
+use App\Transformers\Chat\MessageTransformer;
+use App\Transformers\UserCompactTransformer;
+use Illuminate\Broadcasting\Channel;
+use Illuminate\Contracts\Broadcasting\ShouldBroadcastNow;
+
+class ChatMessageEvent implements ShouldBroadcastNow
```
Comment (nanaya) [inline app/Events/ChatMessageEvent.php:14]: at over half second for 10k users this may not quite be good to be run synchronously as part of the request...

**Review by nanaya — CHANGES_REQUESTED** [review]
not quite done checking :eyes:

**Comment by nanaya** [comment]
Merge conflict...

**Review by nanaya — CHANGES_REQUESTED** [review]

File: app/Transformers/Chat/MessageTransformer.php [hunk 2]
```diff
@@ -13,6 +14,7 @@ class MessageTransformer extends TransformerAbstract
 {
     protected $availableIncludes = [
         'sender',
+        'uuid',
```
[the remark this answers is not in the document]
Comment (notbakaneko) [reply app/Transformers/Chat/MessageTransformer.php:17]: it's not useful for normal listings, though

[the remark this answers is not in the document]
Comment (nanaya) [reply app/Transformers/Chat/MessageTransformer.php:17]: only set if set? :thinking:

File: resources/assets/lib/stores/channel-store.ts [hunk 3]
```diff
@@ -129,60 +128,27 @@ export default class ChannelStore {
...
-
-    try {
-      const response = await this.api.getMessages(channel.channelId, { until });
-      this.handleChatChannelNewMessages(channelId, response);
-    } finally {
-      runInAction(() => {
-        channel.loadingEarlierMessages = false;
-      });
+    if (channel != null) {
+      void channel.loadEarlierMessages();
```
Comment (nanaya) [inline resources/assets/lib/stores/channel-store.ts:151]: looks like can be inlined with some `?.`

**Review by nanaya — CHANGES_REQUESTED** [review]

File: resources/assets/lib/stores/channel-store.ts [hunk 4]
```diff
@@ -213,36 +179,41 @@ export default class ChannelStore {
...
+    if (updateJson.messages != null) {
+      this.updateLastReceivedMessageId(updateJson.messages);
 
-    const groups = groupBy(updateJson.messages, 'channel_id');
-    for (const key of Object.keys(groups)) {
-      const channelId = parseInt(key, 10);
-      this.handleChatChannelNewMessages(channelId, groups[channelId]);
+      const groups = groupBy(updateJson.messages, 'channel_id');
+      for (const key of Object.keys(groups)) {
+        const channelId = parseInt(key, 10);
```
[the remark this answers is not in the document]
Comment (notbakaneko) [reply resources/assets/lib/stores/channel-store.ts:208]: `groupBy` types the key as `string`

[the remark this answers is not in the document]
Comment (nanaya) [reply resources/assets/lib/stores/channel-store.ts:204]: or more like object keys are all string because js object isn't proper map =[

**Review by nanaya — APPROVED** [review]

**Pull request merged**